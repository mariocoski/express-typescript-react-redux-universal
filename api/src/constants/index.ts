export const ONE_HOUR = 3600000;
export const TWO_HOURS = 7200000;