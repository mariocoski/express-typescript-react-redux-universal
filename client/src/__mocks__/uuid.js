const uuid = jest.genMockFromModule('uuid');
uuid.v4 = jest.fn(() => {
  return 1;
});
module.exports = uuid;