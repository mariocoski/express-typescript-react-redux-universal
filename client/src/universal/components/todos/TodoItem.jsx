import React from 'react';
import { List } from 'semantic-ui-react';
const TodoItem = (props) => (
  <List.Item>
      <List.Content>
        <List.Header>{props.todo.title}</List.Header>
        An excellent companion
      </List.Content>
  </List.Item>
);

export default TodoItem;