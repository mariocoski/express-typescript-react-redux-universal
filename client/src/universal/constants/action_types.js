export const ADD_TODO = 'ADD_TODO';
export const REMOVE_TODO = 'REMOVE_TODO';
export const TOGGLE_TODO = 'TOGGLE_TODO';
export const COMPLETE_TODO = 'COMPLETE_TODO';
export const SET_TODOS_FILTER = 'SET_TODOS_FILTER';