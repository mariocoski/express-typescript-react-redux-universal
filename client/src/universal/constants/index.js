export const MOBILE_BREAKPOINT = '768px';
export const TABLET_BREAKPOINT = '991px';
export const SMALL_MONITOR = '1199px';
export const LARGE_MONITOR = '1200px';
export const API_URL = 'http://localhost:3000';
export const CLIENT_URL = 'http://localhost:8080';
export const JWT_TOKEN = 'jwt_token';