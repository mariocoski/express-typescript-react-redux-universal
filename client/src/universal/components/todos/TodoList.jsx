import React from 'react';
import TodoItem from './TodoItem';
import {List, Segment, Grid, Input} from 'semantic-ui-react'
const TodoList = (props) => {

  const todos = props.todos.map((todo,key) => 
    <TodoItem 
      todo={todo} key={key} 
      complete={props.complete} remove={props.remove} 
    >
    </TodoItem>
  );
  
  console.log(props);
  return ( <Grid columns={2} padded>
            <Grid.Column>
              {props.todos.length > 0 && <Segment inverted>
                <List divided relaxed inverted>
                  {todos}
                </List>
              </Segment>
              }
              <Input
                  action={
                      { 
                        color: 'teal', labelPosition: 'right', icon: 'send', 
                        content: 'Add todo', onClick: ()=>{props.add('new one')} 
                      }
                    }
                  defaultValue=''
                  fluid
                  placeholder="your todo..."
                />
             
            </Grid.Column>
      </Grid>);


}

export default TodoList;

