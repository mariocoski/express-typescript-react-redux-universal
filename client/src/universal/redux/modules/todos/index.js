import { ADD_TODO, REMOVE_TODO, TOGGLE_TODO, SET_TODOS_FILTER } from '../../../constants/action_types';
import { SHOW_ALL } from '../../../constants';
import {v4} from 'uuid';

const TODOS_INITIAL_STATE = [];

export function todosReducer(state = TODOS_INITIAL_STATE,action){
  
  switch(action.type){
    case ADD_TODO:
      return [...state,{
        id: v4(),
        title: action.payload,
        completed: false
      }];
    case REMOVE_TODO: 
      return [...state].filter(todo => todo.id !== action.payload);
    case TOGGLE_TODO: 
      const index = state.findIndex(todo => todo.id === action.payload);
      const toggledTodo = {...state[index],completed: ! state[index].completed};
      return [...state].filter(todo => todo.id !== action.payload).concat([toggledTodo]);
  }
  return state;
}

const FILTER_INITIAL_STATE = SHOW_ALL;

export function todosFilterReducer(state = FILTER_INITIAL_STATE, action){
  if(action.type === SET_TODOS_FILTER){
    return action.payload;
  }
  return state;
}

export function setTodosFilter(filter){
  return { 
    type: SET_TODOS_FILTER,
    payload: filter
  }
}

export function addTodo(todo){
  return {
    type: ADD_TODO,
    payload: todo
  }
}

export function removeTodo(id){
  return {
    type: REMOVE_TODO,
    payload: id
  }
}

export function toggleTodo(id){
  return {
    type: TOGGLE_TODO,
    payload: id
  }
}

