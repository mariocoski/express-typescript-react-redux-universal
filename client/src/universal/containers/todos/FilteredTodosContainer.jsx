import React, {Component} from 'react';
import TodoList from '../../components/todos/TodoList';
import {addTodo, removeTodo, toggleTodo, setTodosFilter} from '../../redux/modules/todos';
import {connect} from 'react-redux';

class FilteredTodosContainer extends Component {

  getFilteredTodos(){
    return this.props.todos.filter(todo => {

      if(this.props.filter === 'SHOW_COMPLETED'){
        return todo.completed === true;
      }
      
      return true;
    });
  }

   render(){
     const todos = this.getFilteredTodos();
    return (
      <TodoList
          todos={todos}
          add={this.props.add}
          remove={this.props.remove}
          toggle={this.props.toggle}
        >
      </TodoList>
  )
   }
  
}

const connectStateToProps = state => ({
  filter: state.todosFilter,
  todos : state.todos,
});

const connectDispatchToProps = dispatch => ({
  add : (todo) => dispatch(addTodo(todo)),
  remove : (id) => dispatch(removeTodo(id)),
  toggle : (id) => dispatch(toggleTodo(id)),
  changeFilter : (filter) => dispatch(setTodosFilter(filter)),
});

export default connect(connectStateToProps,connectDispatchToProps)(FilteredTodosContainer);