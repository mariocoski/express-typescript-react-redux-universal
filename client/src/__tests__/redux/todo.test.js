import { addTodo, removeTodo, toggleTodo, completeTodo,todosReducer,todosFilterReducer, setTodosFilter } from '../../universal/redux/modules/todos/index';
import { ADD_TODO, REMOVE_TODO, TOGGLE_TODO, SET_TODOS_FILTER } from '../../universal/constants/action_types';
import {v4} from 'uuid';
import React from 'react';
import renderer from 'react-test-renderer';

describe('REDUX', () => {

  describe('TODOS', () => {

    it('Should be able to create add todo action', () => {
      const text = 'buy milk';
      const expectedAction = {
        type: ADD_TODO,
        payload: text
      };
      expect(addTodo(text)).toEqual(expectedAction);
    });

    it('Should be able to create remove todo action', () => {
      const id = 1;
      const expectedAction = {
        type: REMOVE_TODO,
        payload: id
      };
      expect(removeTodo(id)).toEqual(expectedAction);
    });

    
    it('Should be able to create toggle action', () => {
      const id = 1;
      const expectedAction = {
        type: TOGGLE_TODO,
        payload: id
      };
      expect(toggleTodo(id)).toEqual(expectedAction);
    });

    it('Reducer should return initial state when no action provided', () => {
      expect(todosReducer(undefined, {})).toEqual([]);
    });

    it('Should handle add todo action', () => {
      const text = 'buy milk';
      const expectedState = [{
        id: v4(),
        title: text,
        completed: false,
      }];
      expect(todosReducer([], addTodo(text))).toEqual(expectedState);
    });

    it('Should handle remove todo action', () => {
      const id = v4();
      const initialState = [{
        id,
        title: 'buy milk',
        completed: false,
      }];
      expect(todosReducer(initialState, removeTodo(id))).toEqual([]);
    });

    it('Should handle toggle todo action', () => {
      const id = v4();
      const text = 'buy butter';
      const initialState = [{
        id,
        title: text,
        completed: false,
      }];

      const expectedState = [{
          id,
          title: text,
          completed: true,
      }];
      expect(todosReducer(initialState, toggleTodo(id))).toEqual(expectedState);
    });
  });

  describe('FILTERS', () => {
    it('Reducer should return initial state when not action provided', () => {

      expect(todosFilterReducer(undefined, {})).toEqual('SHOW_ALL');
    });

    it('Should be able to create set filter action', () => {
      const text = 'SHOW_COMPLETED';
      const expectedAction = {
        type: SET_TODOS_FILTER,
        payload: text
      };
      expect(setTodosFilter(text)).toEqual(expectedAction);
    });

    it('Should handle set todos filter action', () => {
      const text = "SHOW_COMPLETED";
      const expectedState = text;
      expect(todosFilterReducer(undefined, setTodosFilter(text))).toEqual(expectedState);
    });

    it('renders correctly', () => {
      const tree = renderer.create(
        <div>OK</div>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });


  });
});