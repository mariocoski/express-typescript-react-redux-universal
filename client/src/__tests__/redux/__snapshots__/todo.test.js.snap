// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`REDUX FILTERS renders correctly 1`] = `
<div>
  OK
</div>
`;
